icicle_v3/backend/cpu/include/tasks_manager.h

```diff
@@ -0,0 +1,309 @@
+#pragma once 
+#include <atomic>
+#include <thread> // TODO check windows support
+#include <stdexcept>
+
+#define LOG_TASKS_PER_THREAD 2
+#define TASKS_PER_THREAD (1 << LOG_TASKS_PER_THREAD)
+#define TASK_IDX_MASK (TASKS_PER_THREAD - 1)
+#define MANAGER_SLEEP_USEC 10
+#define THREAD_SLEEP_USEC 1
+
+/**
+ * @class TasksManager
+ * @brief Class for managing parallel executions of small `Task`s which are child class of `TaskBase` described bellow.
+ * 
+ * The class manages a vector of `Worker`s, which are threads and additional required data members for executions of 
+ * `Task`s. The `Task`s are split in a thread-pool fashion - finding free slot in the `Worker`s for the user to set up 
+ * additional tasks, and fetching completed `Task`s back to the user.
+ */
+template<class Task>
+class TasksManager {
+public:
+  /**
+   * @brief Constructor for `TaskManager`.
+   * @param nof_workers - number of workers / threads to be ran simultaneously
+   */
+  TasksManager(const int nof_workers) : workers(nof_workers) {}
+
+  /**
+   * @brief Get free slot to insert new task to be executed. This is a blocking function - until a free task is found.
+   * @param task_slot - pointer to the internal task to allow the user to edit in the new task.
+   * @return boolean value indication of the current content of the task - true indicates that `task_slot` holds a 
+   * previous result that should be handled by the user before writing the new task.
+   */
+  bool get_free_task(Task*& task_slot);
```

Comment:
I don't understand. If it's a free task then why do I need to handle a completed task first? 
I would expect to it to fail and so I would have to call the get_completed_task() and handle it. Maybe there you can modify the completed one and redispatch it.

Reply:
the behavior of the function is correct. i,e, this is what needed.
the name of the function can change to be more clear e.g. get_free_or_completed_task

Reply:
the code marked the task returned as if its result has been handled for it to not give it to the user again in the case of 2 consecutive calls to get_free_task
The main issue is for completed task but free task is implemented similarly

I can add another state - HAS_RESULT_BUT_ALREADY_SENT_TO_USER but it is a bit cumbersome